File: activities/creature-battle/csharp/creature-battle-solution/Goblin.cs
namespace creature_battle_solution;

using System;

public class Goblin : Creature
{
    // --- Properties ---
    public double DodgeChance { get; private set; }
    private Random _random = new Random();

    // --- Constructor ---
    public Goblin(string name, int health, int attackPower, double dodgeChance)
        : base(name, health, attackPower)
    {
        // Ensure dodge chance is between 0.0 and 1.0
        DodgeChance = Math.Clamp(dodgeChance, 0.0, 1.0);
    }

    // --- Overridden Method ---

    /// <summary>
    /// Overrides the base TakeDamage method. The Goblin has a chance
    /// to dodge the attack completely.
    /// </summary>
    public override void TakeDamage(int damage)
    {
        if (_random.NextDouble() < DodgeChance)
        {
            Console.WriteLine($"{Name} dodges the attack!");
        }
        else
        {
            // If the dodge fails, call the original base method
            base.TakeDamage(damage);
        }
    }
}

File: activities/creature-battle/java/creature-battle-solution/src/main/java/org/example/Goblin.java
package org.example;

public class Goblin extends Creature {
    private double dodgeChance;

    public Goblin(String name, int health, int attackPower, double dodgeChance) {
        super(name, health, attackPower);
        this.dodgeChance = dodgeChance;
    }

    // Method Override
    @Override
    public void takeDamage(int damage) {
        if (Math.random() < this.dodgeChance) {
            System.out.println(this.name + " dodged the attack!");
        } else {
            super.takeDamage(damage); // Let the parent class handle the damage
        }
    }
}